batcher-bot/liquidity/src/bot.ts

```diff
@@ -0,0 +1,231 @@
+                buy_token,
+                sell_token,
+                details
+              );
+
+              if (order_list_opt.isSome()) {
+                const ords = order_list_opt.get();
+                for (let j = 0; j < ords.length; j++) {
+                  let ord = ords[i];
+                  console.info("Provisioning -> ", ord);
+                  submit_deposit(details, ord, tezos).then(() => {
+                    console.info("Provisioned order.");
+                  });
+                }
+              }
+            }
+          }
+        }
+      }
+    } catch (error: any) {
+      console.error(error);
+    }
+  }
+};
+
+const jit_provision = (
+  tezos: TezosToolkit,
+  message: any,
+  details: contract_details,
+  settings: liquidity_settings
+) => {
+  for (let i = 0; i < Object.keys(message.data).length; i++) {
+    try {
+      const msg = message.data[i];
+      if (msg.parameter) {
+        const entrypoint = msg.parameter.entrypoint;
+        const sender = msg.sender.address;
+        if (entrypoint == "deposit" && sender != details.user_address) {
+          console.info("Deposit msg", msg);
+          const val = msg.parameter.value;
+          const pair: string = getPairName(
+            val.swap.from.token.name,
+            val.swap.to.name
+          );
+
+          const current_batch_indices =
+            msg.storage.batch_set.current_batch_indices;
+          const batch_number = current_batch_indices[pair];
+
+          if (settings.token_pairs.has(pair)) {
+            const jit_setting = settings.token_pairs.get(pair);
+            if (jit_setting) {
+              const order = parse_deposit(val);
+              const order_opt = can_provision_jit(
+                batch_number,
+                jit_setting,
+                order
+              );
+              if (order_opt.isSome()) {
+                const ord = order_opt.get();
+                console.info("Provisioning -> ", JSON.stringify(ord));
+                submit_deposit(details, ord, tezos).then(() => {
+                  console.info("Provisioned order.");
+                });
+              }
+            }
+          }
+        }
+      }
+    } catch (error: any) {
+      console.error(error);
+    }
+  }
+};
+
+export const run_jit = async (
+  tezos: TezosToolkit,
+  details: contract_details,
+  settings: liquidity_settings,
+  socket_connection: HubConnection
+) => {
+  await socket_connection.start();
+
+  await socket_connection.invoke("SubscribeToOperations", {
+    address: details.address,
+    types: "transaction",
+  });
+
+  await socket_connection.invoke("SubscribeToBigMaps", {
+    contract: details.address,
+  });
+
+  socket_connection.on("operations", (msg: any) => {
+    if (!msg.data) return;
+    jit_provision(tezos, msg, details, settings);
+  });
+
+  socket_connection.on("bigmaps", (msg: any) => {
+    if (!msg.data) return;
+    redeem_on_cleared(details.address, msg, tezos);
+  });
```

Comment:
Don't you want to declare the events before invoking the subscribe operations? Imo it would be more safe in case an event happens between the `invoke` and the `on`

Reply:
Fixed